# **Capstone III Data Wrangling**

### **Section 1: Imports and Initial Inspection**

In [1]:
import pandas as pd

In [2]:
train = pd.read_csv("../Raw Data/train.csv")
test = pd.read_csv('../Raw Data/test.csv')

In [3]:
train.head()

Id        Date   Time  CO2_(dinning-room)  CO2_room  \
0   0  13/03/2012  11:45             216.560   221.920   
1   1  13/03/2012  12:00             219.947   220.363   
2   2  13/03/2012  12:15             219.403   218.933   
3   3  13/03/2012  12:30             218.613   217.045   
4   4  13/03/2012  12:45             217.714   216.080   

   Relative_humidity_(dinning-room)  Relative_humidity_room  \
0                           39.9125                 42.4150   
1                           39.9267                 42.2453   
2                           39.7720                 42.2267   
3                           39.7760                 42.0987   
4                           39.7757                 42.0686   

   Lighting_(dinning-room)  Lighting_room  Meteo_Rain  Meteo_Sun_dusk  \
0                  81.6650        113.520         0.0         623.360   
1                  81.7413        113.605         0.0         623.211   
2                  81.4240        113.600         0.0         622.656   
3                  81.5013        113.344         0.0         622.571   
4                  81.4657        113.034         0.0         622.400   

   Meteo_Wind  Meteo_Sun_light_in_west_facade  Meteo_Sun_light_in_east_facade  \
0     1.42625                         9690.24                        12604.20   
1     1.59200                        11022.00                        10787.20   
2     1.89133                        13960.50                         9669.63   
3     1.82800                        18511.20                         9648.13   
4     2.36071                        26349.00                         9208.32   

   Meteo_Sun_light_in_south_facade  Meteo_Sun_irradiance  \
0                          95436.8               758.880   
1                          95436.8               762.069   
2                          95398.6               766.251   
3                          95360.3               766.037   
4                          95354.9               762.743   

   Outdoor_relative_humidity_Sensor  Day_of_the_week  Indoor_temperature_room  
0                           48.3750              2.0                  17.8275  
1                           47.8080              2.0                  18.1207  
2                           47.4320              2.0                  18.4367  
3                           47.0240              2.0                  18.7513  
4                           45.4743              2.0                  19.0414

In [4]:
test.head()

Id        Date   Time  CO2_(dinning-room)  CO2_room  \
0  2764  18/04/2012  00:00             207.840   206.560   
1  2765  18/04/2012  00:15             207.499   206.101   
2  2766  18/04/2012  00:30             207.360   206.805   
3  2767  18/04/2012  00:45             207.445   206.325   
4  2768  18/04/2012  01:00             207.840   205.952   

   Relative_humidity_(dinning-room)  Relative_humidity_room  \
0                           31.5800                 30.1600   
1                           31.6240                 30.0613   
2                           31.6507                 30.1773   
3                           31.6987                 30.2560   
4                           31.7707                 30.3840   

   Lighting_(dinning-room)  Lighting_room  Meteo_Rain  Meteo_Sun_dusk  \
0                  11.7200        14.1600         0.0            0.65   
1                  11.5240        14.7600         0.0            0.65   
2                  11.6220        14.9547         0.0            0.65   
3                  11.7847        15.0520         0.0            0.65   
4                  11.5887        15.2147         0.0            0.65   

   Meteo_Wind  Meteo_Sun_light_in_west_facade  Meteo_Sun_light_in_east_facade  \
0    0.000000                             0.0                             0.0   
1    0.000000                             0.0                             0.0   
2    0.088000                             0.0                             0.0   
3    0.127333                             0.0                             0.0   
4    0.000000                             0.0                             0.0   

   Meteo_Sun_light_in_south_facade  Meteo_Sun_irradiance  \
0                              0.0              -3.12000   
1                              0.0              -2.80800   
2                              0.0              -2.96400   
3                              0.0              -3.02933   
4                              0.0              -2.88600   

   Outdoor_relative_humidity_Sensor  Day_of_the_week  
0                           39.0800              3.0  
1                           38.9680              3.0  
2                           38.9467              3.0  
3                           39.0760              3.0  
4                           40.0413              3.0

In [5]:
train.shape

(2764, 19)

In [6]:
test.shape

(1373, 18)

In [7]:
# With the exception of Date and Time, all columns are the correct data type
train.dtypes

Id                                    int64
Date                                 object
Time                                 object
CO2_(dinning-room)                  float64
CO2_room                            float64
Relative_humidity_(dinning-room)    float64
Relative_humidity_room              float64
Lighting_(dinning-room)             float64
Lighting_room                       float64
Meteo_Rain                          float64
Meteo_Sun_dusk                      float64
Meteo_Wind                          float64
Meteo_Sun_light_in_west_facade      float64
Meteo_Sun_light_in_east_facade      float64
Meteo_Sun_light_in_south_facade     float64
Meteo_Sun_irradiance                float64
Outdoor_relative_humidity_Sensor    float64
Day_of_the_week                     float64
Indoor_temperature_room             float64
dtype: object

In [8]:
test.dtypes

Id                                    int64
Date                                 object
Time                                 object
CO2_(dinning-room)                  float64
CO2_room                            float64
Relative_humidity_(dinning-room)    float64
Relative_humidity_room              float64
Lighting_(dinning-room)             float64
Lighting_room                       float64
Meteo_Rain                          float64
Meteo_Sun_dusk                      float64
Meteo_Wind                          float64
Meteo_Sun_light_in_west_facade      float64
Meteo_Sun_light_in_east_facade      float64
Meteo_Sun_light_in_south_facade     float64
Meteo_Sun_irradiance                float64
Outdoor_relative_humidity_Sensor    float64
Day_of_the_week                     float64
dtype: object

In [9]:
# No NaN values
train.isna().sum()

Id                                  0
Date                                0
Time                                0
CO2_(dinning-room)                  0
CO2_room                            0
Relative_humidity_(dinning-room)    0
Relative_humidity_room              0
Lighting_(dinning-room)             0
Lighting_room                       0
Meteo_Rain                          0
Meteo_Sun_dusk                      0
Meteo_Wind                          0
Meteo_Sun_light_in_west_facade      0
Meteo_Sun_light_in_east_facade      0
Meteo_Sun_light_in_south_facade     0
Meteo_Sun_irradiance                0
Outdoor_relative_humidity_Sensor    0
Day_of_the_week                     0
Indoor_temperature_room             0
dtype: int64

In [10]:
test.isna().sum()

Id                                  0
Date                                0
Time                                0
CO2_(dinning-room)                  0
CO2_room                            0
Relative_humidity_(dinning-room)    0
Relative_humidity_room              0
Lighting_(dinning-room)             0
Lighting_room                       0
Meteo_Rain                          0
Meteo_Sun_dusk                      0
Meteo_Wind                          0
Meteo_Sun_light_in_west_facade      0
Meteo_Sun_light_in_east_facade      0
Meteo_Sun_light_in_south_facade     0
Meteo_Sun_irradiance                0
Outdoor_relative_humidity_Sensor    0
Day_of_the_week                     0
dtype: int64

### **Section 2: Format and Structure Changes**

Here we'll make some format changes. This includes creating a combined date and time column, changing the data types of the date columns, and changing column names.

In [11]:
train.head()

Id        Date   Time  CO2_(dinning-room)  CO2_room  \
0   0  13/03/2012  11:45             216.560   221.920   
1   1  13/03/2012  12:00             219.947   220.363   
2   2  13/03/2012  12:15             219.403   218.933   
3   3  13/03/2012  12:30             218.613   217.045   
4   4  13/03/2012  12:45             217.714   216.080   

   Relative_humidity_(dinning-room)  Relative_humidity_room  \
0                           39.9125                 42.4150   
1                           39.9267                 42.2453   
2                           39.7720                 42.2267   
3                           39.7760                 42.0987   
4                           39.7757                 42.0686   

   Lighting_(dinning-room)  Lighting_room  Meteo_Rain  Meteo_Sun_dusk  \
0                  81.6650        113.520         0.0         623.360   
1                  81.7413        113.605         0.0         623.211   
2                  81.4240        113.600         0.0         622.656   
3                  81.5013        113.344         0.0         622.571   
4                  81.4657        113.034         0.0         622.400   

   Meteo_Wind  Meteo_Sun_light_in_west_facade  Meteo_Sun_light_in_east_facade  \
0     1.42625                         9690.24                        12604.20   
1     1.59200                        11022.00                        10787.20   
2     1.89133                        13960.50                         9669.63   
3     1.82800                        18511.20                         9648.13   
4     2.36071                        26349.00                         9208.32   

   Meteo_Sun_light_in_south_facade  Meteo_Sun_irradiance  \
0                          95436.8               758.880   
1                          95436.8               762.069   
2                          95398.6               766.251   
3                          95360.3               766.037   
4                          95354.9               762.743   

   Outdoor_relative_humidity_Sensor  Day_of_the_week  Indoor_temperature_room  
0                           48.3750              2.0                  17.8275  
1                           47.8080              2.0                  18.1207  
2                           47.4320              2.0                  18.4367  
3                           47.0240              2.0                  18.7513  
4                           45.4743              2.0                  19.0414

In [12]:
test.head()

Id        Date   Time  CO2_(dinning-room)  CO2_room  \
0  2764  18/04/2012  00:00             207.840   206.560   
1  2765  18/04/2012  00:15             207.499   206.101   
2  2766  18/04/2012  00:30             207.360   206.805   
3  2767  18/04/2012  00:45             207.445   206.325   
4  2768  18/04/2012  01:00             207.840   205.952   

   Relative_humidity_(dinning-room)  Relative_humidity_room  \
0                           31.5800                 30.1600   
1                           31.6240                 30.0613   
2                           31.6507                 30.1773   
3                           31.6987                 30.2560   
4                           31.7707                 30.3840   

   Lighting_(dinning-room)  Lighting_room  Meteo_Rain  Meteo_Sun_dusk  \
0                  11.7200        14.1600         0.0            0.65   
1                  11.5240        14.7600         0.0            0.65   
2                  11.6220        14.9547         0.0            0.65   
3                  11.7847        15.0520         0.0            0.65   
4                  11.5887        15.2147         0.0            0.65   

   Meteo_Wind  Meteo_Sun_light_in_west_facade  Meteo_Sun_light_in_east_facade  \
0    0.000000                             0.0                             0.0   
1    0.000000                             0.0                             0.0   
2    0.088000                             0.0                             0.0   
3    0.127333                             0.0                             0.0   
4    0.000000                             0.0                             0.0   

   Meteo_Sun_light_in_south_facade  Meteo_Sun_irradiance  \
0                              0.0              -3.12000   
1                              0.0              -2.80800   
2                              0.0              -2.96400   
3                              0.0              -3.02933   
4                              0.0              -2.88600   

   Outdoor_relative_humidity_Sensor  Day_of_the_week  
0                           39.0800              3.0  
1                           38.9680              3.0  
2                           38.9467              3.0  
3                           39.0760              3.0  
4                           40.0413              3.0

In [13]:
# Drop the Id columns. We'll be indexing our dataframe by date/time.
train.drop(columns = 'Id', inplace = True)
test.drop(columns = 'Id', inplace = True)

In [14]:
# We'd like these column names to be formatted differently.
train.columns

Index(['Date', 'Time', 'CO2_(dinning-room)', 'CO2_room',
       'Relative_humidity_(dinning-room)', 'Relative_humidity_room',
       'Lighting_(dinning-room)', 'Lighting_room', 'Meteo_Rain',
       'Meteo_Sun_dusk', 'Meteo_Wind', 'Meteo_Sun_light_in_west_facade',
       'Meteo_Sun_light_in_east_facade', 'Meteo_Sun_light_in_south_facade',
       'Meteo_Sun_irradiance', 'Outdoor_relative_humidity_Sensor',
       'Day_of_the_week', 'Indoor_temperature_room'],
      dtype='object')

In [15]:
test.columns

Index(['Date', 'Time', 'CO2_(dinning-room)', 'CO2_room',
       'Relative_humidity_(dinning-room)', 'Relative_humidity_room',
       'Lighting_(dinning-room)', 'Lighting_room', 'Meteo_Rain',
       'Meteo_Sun_dusk', 'Meteo_Wind', 'Meteo_Sun_light_in_west_facade',
       'Meteo_Sun_light_in_east_facade', 'Meteo_Sun_light_in_south_facade',
       'Meteo_Sun_irradiance', 'Outdoor_relative_humidity_Sensor',
       'Day_of_the_week'],
      dtype='object')

In [20]:
# Function to reformat column names
def column_renamer(df):
    new_names = []
    for column in df.columns:
        new_name = column.lower() # Make column names lower case
        
        new_name = (new_name.replace('(', '').replace(')', '').replace('-', '_')
        .replace('dinning', 'dining').replace('meteo_', '').replace('_sensor', '')) # Replaces parentheses, misspellings, etc.
        
        new_names.append(new_name)
        mapper = dict(zip(df.columns, new_names))
        df.rename(mapper, axis = 'columns', inplace = True)
    return(df)

In [24]:
column_renamer(train)
train.columns

Index(['date', 'time', 'co2_dining_room', 'co2_room',
       'relative_humidity_dining_room', 'relative_humidity_room',
       'lighting_dining_room', 'lighting_room', 'rain', 'sun_dusk', 'wind',
       'sun_light_in_west_facade', 'sun_light_in_east_facade',
       'sun_light_in_south_facade', 'sun_irradiance',
       'outdoor_relative_humidity', 'day_of_the_week',
       'indoor_temperature_room'],
      dtype='object')

In [25]:
column_renamer(test)
test.columns

Index(['date', 'time', 'co2_dining_room', 'co2_room',
       'relative_humidity_dining_room', 'relative_humidity_room',
       'lighting_dining_room', 'lighting_room', 'rain', 'sun_dusk', 'wind',
       'sun_light_in_west_facade', 'sun_light_in_east_facade',
       'sun_light_in_south_facade', 'sun_irradiance',
       'outdoor_relative_humidity', 'day_of_the_week'],
      dtype='object')

In [26]:
# Add a new column that combines the Date and Time columns. Drop the time column.
train['date_time'] = train['date'] + ' ' + train['time']
train.drop(columns = 'time', inplace = True)

test['date_time'] = test['date'] + ' ' + test['time']
test.drop(columns = 'time', inplace = True)

train['date_time'].sample(10)

336     16/03/2012 23:45
2447    07/04/2012 23:30
1313    27/03/2012 04:00
1131    25/03/2012 06:30
2248    05/04/2012 21:45
2297    06/04/2012 10:00
2122    04/04/2012 14:15
766     21/03/2012 11:15
811     21/03/2012 22:30
454     18/03/2012 05:15
Name: date_time, dtype: object

In [27]:
test['date_time'].sample(10)

774     26/04/2012 01:30
808     26/04/2012 10:00
1315    01/05/2012 16:45
560     23/04/2012 20:00
398     22/04/2012 03:30
809     26/04/2012 10:15
1316    01/05/2012 17:00
783     26/04/2012 03:45
125     19/04/2012 07:15
43      18/04/2012 10:45
Name: date_time, dtype: object

In [28]:
# These should be datetime objects.
train[['date', 'date_time']].dtypes, test[['date', 'date_time']].dtypes

(date         object
 date_time    object
 dtype: object,
 date         object
 date_time    object
 dtype: object)

In [29]:
# Change Date and Date_time columns to be datetime objects instead of strings.
train['date'] = pd.to_datetime(train['date'], format = '%d/%m/%Y')
train['date_time'] = pd.to_datetime(train['date_time'], format = "%d/%m/%Y %H:%M")

test['date'] = pd.to_datetime(test['date'], format = '%d/%m/%Y')
test['date_time'] = pd.to_datetime(test['date_time'], format = "%d/%m/%Y %H:%M")

train[['date', 'date_time']].sample(10)

date           date_time
1268 2012-03-26 2012-03-26 16:45:00
1446 2012-03-28 2012-03-28 13:15:00
828  2012-03-22 2012-03-22 02:45:00
1646 2012-03-30 2012-03-30 15:15:00
44   2012-03-13 2012-03-13 22:45:00
702  2012-03-20 2012-03-20 19:15:00
476  2012-03-18 2012-03-18 10:45:00
2082 2012-04-04 2012-04-04 04:15:00
2340 2012-04-06 2012-04-06 20:45:00
2609 2012-04-09 2012-04-09 16:00:00

In [30]:
test[['date', 'date_time']].sample(10)

date           date_time
30  2012-04-18 2012-04-18 07:30:00
871 2012-04-27 2012-04-27 01:45:00
912 2012-04-27 2012-04-27 12:00:00
146 2012-04-19 2012-04-19 12:30:00
712 2012-04-25 2012-04-25 10:00:00
539 2012-04-23 2012-04-23 14:45:00
918 2012-04-27 2012-04-27 13:30:00
460 2012-04-22 2012-04-22 19:00:00
93  2012-04-18 2012-04-18 23:15:00
649 2012-04-24 2012-04-24 18:15:00

In [31]:
train.dtypes, test.dtypes

(date                             datetime64[ns]
 co2_dining_room                         float64
 co2_room                                float64
 relative_humidity_dining_room           float64
 relative_humidity_room                  float64
 lighting_dining_room                    float64
 lighting_room                           float64
 rain                                    float64
 sun_dusk                                float64
 wind                                    float64
 sun_light_in_west_facade                float64
 sun_light_in_east_facade                float64
 sun_light_in_south_facade               float64
 sun_irradiance                          float64
 outdoor_relative_humidity               float64
 day_of_the_week                         float64
 indoor_temperature_room                 float64
 date_time                        datetime64[ns]
 dtype: object,
 date                             datetime64[ns]
 co2_dining_room                         float64
 co2

### **Section 3: Data Validation**

Now we'll inspect the contents of our columns to make sure the data makes sense. We know there are no missing values, but we need to make sure the data isn't invalid in other ways. For example, we shouldn't see negative values in columns like the co2 columns, which are in parts per million. We also need to see if any measurements were skipped. They should occur every fifteen minutes.

In [32]:
# Measurements are supposed to have been taken every 15 minutes. Let's find out if any rows are missing.
# .diff() takes the difference between each row and the previous row.
train['date_time'].diff().drop([0]).max(), train['date_time'].diff().drop([0]).min()

(Timedelta('0 days 00:15:00'), Timedelta('0 days 00:15:00'))

In [33]:
test['date_time'].diff().drop([0]).max(), test['date_time'].diff().drop([0]).min()

(Timedelta('0 days 00:30:00'), Timedelta('0 days 00:15:00'))

Train is fine, meaning the maximum and minimum difference in time between any two subsequent measurments is 15 minutes. In the test set it looks like there is a maximum difference of 30 minutes, meaning at least one measurement was likely skipped. We must investigate this.

In [34]:
# There are two instances of a gap of 30 minutes between measurements
test['date_time'].diff().drop([0]).sort_values(ascending = False)

1362   0 days 00:30:00
1356   0 days 00:30:00
912    0 days 00:15:00
920    0 days 00:15:00
919    0 days 00:15:00
             ...      
456    0 days 00:15:00
455    0 days 00:15:00
454    0 days 00:15:00
453    0 days 00:15:00
1372   0 days 00:15:00
Name: date_time, Length: 1372, dtype: timedelta64[ns]

In [35]:
# Our culprits both occur on May 2, from 02:45 to 3:15, and again from 4:30 to 5:00.
test['date_time'].iloc[1355:1363]

1355   2012-05-02 02:45:00
1356   2012-05-02 03:15:00
1357   2012-05-02 03:30:00
1358   2012-05-02 03:45:00
1359   2012-05-02 04:00:00
1360   2012-05-02 04:15:00
1361   2012-05-02 04:30:00
1362   2012-05-02 05:00:00
Name: date_time, dtype: datetime64[ns]

There are different strategies for imputing missing data. In non time series data sets, common strategies include imputing the mean or median values for the missing values. For a time series data set like ours, strategies include duplicating the last value, duplicating the next value, linear interpolation, and 'spline' interpolation, where the goal is to minimize the curvature of the imputed data point. Another strategy common to all data sets is to simply drop the missing values. In our current situation, we're only missing 2 out of 1373 data points. Our intutions about weather and temperature lead us to believe that it's very unlikely to be suprised by radical deviations in fifteen minute or half hour intervals (in other words, our data has high autocorrelation which we'll see in the EDA section). Therefore, if this data was missing in the sense of NaN values or invalid values, we would drop those two rows. The rows have already been dropped for us by the uploaders. We will not attempt to impute any values.

In [36]:
# Function to check if the minimum values of the numeric data types are above zero
def min_max_checker(df):
    invalid_value_columns = []
    for column in df.columns:
        try:
            df[column][0] + 1 # Checks to see if column is of numeric type
            if df[column].min() < 0: # If it is numeric, and if the minimum value is below zero, add it to the list
                invalid_value_columns.append(column)
        except:
            pass # Does nothing if the column is not numeric
    return(invalid_value_columns) # Return a list of columns with minimum values below zero.

In [37]:
min_max_checker(train), min_max_checker(test)

(['sun_irradiance'], ['sun_irradiance'])

In [38]:
train['sun_irradiance'].describe()

count    2764.000000
mean      215.010017
std       297.234046
min        -4.164670
25%        -3.381330
50%         3.922000
75%       435.434500
max      1028.270000
Name: sun_irradiance, dtype: float64

In [39]:
test['sun_irradiance'].describe()

count    1373.000000
mean      266.816322
std       338.532417
min        -3.708000
25%        -3.042000
50%        35.816000
75%       566.293000
max      1094.660000
Name: sun_irradiance, dtype: float64

In [40]:
# At least 25% of irradiance values are negative. Let's see exactly what proportion of irradiance's values are negative.
negative_irrad_rows = len(train[train['sun_irradiance'] < 0])
all_rows = len(train)
print(round(negative_irrad_rows / all_rows * 100, 2), 'percent of the training irradiance readings are below zero.')

49.13 percent of the training irradiance readings are below zero.


In [41]:
negative_irrad_rows = len(test[test['sun_irradiance'] < 0])
all_rows = len(test)
print(round(negative_irrad_rows / all_rows * 100, 2), 'percent of the testing irradiance readings are below zero.')

44.28 percent of the testing irradiance readings are below zero.


Upon cursory investigation, it seems to be relatively common to have small negative irradiance values. The reason for this might be because the value is recorded on a log scale, with an arbitrary zero point at, say, dawn or dusk. At night (which comprises roughly half of the measurments), the absence of sun falls results in low points of irradiance that fall just below the arbitrary zero point, meaning the value of the exponent variable is negative. During the day, it is vastly higher.

Posts that lead to this conclusion:

https://unmethours.com/question/13731/negative-and-positive-values-of-solar-radiation-during-night/
https://www.researchgate.net/post/Its_normal_to_have_negative_values_of_irradiance_in_a_database

In [42]:
# Lastly, we'll set the indices of the train and test sets to be our date_time columns.
train.set_index('date_time', inplace = True)
test.set_index('date_time', inplace = True)

In [43]:
# Save cleaned version of the data.
train.to_csv('../Data/train_cleaned.csv')
test.to_csv('../Data/test_cleaned.csv')

# Summary

Here is a summary of the changes we have made to the data so far:

1) Made combined date_time columns out of the date and time columns,

2) Changed the data types of the date and date_time columns to be datetime objects,

3) Deleted the time and Id columns,

4) Set the indices of our dataframes to be the combined date_time columns,

5) Fixed spelling errors and in general made the column name formatting consistent and less verbose.


The data was quite clean when it came to us. There were no missing values in the form of NaN's we needed to deal with. We further confirmed that the columns weren't missing values in other ways by observing the maximum and minimum values of the differences between each row. Train had no more and no less than a 15 minute difference between two subsequent rows, and test only had two instances 30 minutes between two subsequent rows (two missing values). These were safe to leave be. Lastly, we made sure that no column had negative values that wasn't supposed to have negative values. 